include/xtensor-io/xgzip.hpp

```diff
@@ -0,0 +1,213 @@
+/***************************************************************************
+* Copyright (c) Wolf Vollprecht, Sylvain Corlay and Johan Mabille          *
+* Copyright (c) QuantStack                                                 *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XTENSOR_IO_GZIP_HPP
+#define XTENSOR_IO_GZIP_HPP
+
+#include <fstream>
+
+#include "zlib.h"
+
+#include "xtensor/xadapt.hpp"
+
+#define GZIP_CHUNK 0x4000
+#define GZIP_WINDOWBITS 15
+#define GZIP_ENCODING 16
+#define ENABLE_ZLIB_GZIP 32
```

Comment:
Not a big expert here, only used a bit of gzip with HDF5. However, It might make sense to allow the user to customise if wanted?

Reply:
Yes, you're right. I'm not an expert of gzip either, most of this code comes from [here](https://www.lemoda.net/c/zlib-open-read/) and [there](https://www.lemoda.net/c/zlib-open-write/index.html).
I tried using https://github.com/mateidavid/zstr (that we already vendor), but without success, and it doesn't store headers. It doesn't allow to set these parameters either.

Reply:
When I did all kinds of downstream additions of HighFive I ended up having a separate options class : https://github.com/BlueBrain/HighFive/blob/40f7624b37e1386fe1ac1885beb474e9346311d6/include/highfive/H5Easy.hpp#L112 . But maybe that's a bit of an overkill for now, and more something to do when someone asks for it.

Reply:
Yes, maybe we could add these options in the future (with default values), but I guess they would go in the `xgzip_config` struct.